### PR TITLE
one-step kernel install

### PR DESCRIPTION
**Pull request by ddollar**
* Convox master account runs `cmd/bootstrap` as a Lambda function
* Convox master account has an SNS topic that can be posted by anyone, linked to this Lambda function
* Kernel boots now attempt to create a lambda function `KernelFormation` using the Convox master SNS as a `ServiceToken`. The CF stack will create credentials (`KernelUser`/`KernelAccess`) and pass them as properties.
* `cmd/bootstrap` in the master account will use the credentials to create the requested lambda function in the user's account
* The ARN for that function is passed back as the physical ID to the requesting lambda stack
* The ARN for that function is used as the `ServiceToken` for all further custom requests in the stack.

### CODE REVIEW

File: cmd/bootstrap/bin/create [hunk 1]
```diff
@@ -0,0 +1,13 @@
+role=$(aws iam create-role --role-name convox-formation --assume-role-policy-document file://data/role-assume.json | jq -r .Role.Arn)
+
+aws iam put-role-policy --role-name convox-formation --policy-name all --policy-document file://data/role-policy.json
```
Comment (nzoschke) [inline cmd/bootstrap/bin/create]: I now suspect my lambda problems in testing were around recreating roles, and not getting the role-policy applied to the latest one.

Just a guess, but perhaps moving the sleep after the create-role and before the put-role-policy could help. The reference by name and not ARN is suspicious.

But this guess will be easy to verify next time I see lambda logging problems.

Comment (ddollar) [inline cmd/bootstrap/bin/create]: Hopefully this shouldn't be a problem any more. This `make create` is designed to be run only once, and only on our master Convox account. Everyone else will use SNS to talk to this guy in order to bootstrap a named lambda function in their own account that uses a CF-generated role.

File: cmd/bootstrap/lambda.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
+exports.external = function(event, context) {
+
+  process.on('uncaughtException', function(err) {
+    return context.done(err);
```
Comment (nzoschke) [inline cmd/bootstrap/lambda.js]: In this patch I'm attempting to add JS that will put a response even if this uncaught exception happens. I have gotten the uncaught exception a fair amount during testing and development, leaving stacks wedged.

https://github.com/convox/kernel/commit/d33c9ea1fb7c495d77a602a97552279ae0e6ffb6

This is inspired by an AWS example:

https://s3.amazonaws.com/cloudformation-examples/lambda/LookupStackOutputs.js

My patch isn't quite working yet however...

File: cmd/bootstrap/lambda.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
+exports.external = function(event, context) {
+
```
Comment (nzoschke) [inline cmd/bootstrap/lambda.js]: We should log the event and context in JS for better debugging.

File: cmd/bootstrap/lambda.js (hunk 2 above)
Comment (ddollar) [inline cmd/bootstrap/lambda.js]: That will be a nice addition

File: dist/staging.json [hunk 4]
```diff
@@ -23,16 +27,16 @@
     "Ami": {
       "Type" : "String",
       "Default" : "ami-d0b9acb8",
-      "Description" : "ECS Cluster AMI (do not change unless you know what you're doing)"
+      "Description" : "ECS Cluster AMI (only experts should change this)"
```
Comment (nzoschke) [inline dist/staging.json]: A nice to have would be automatic additional region support. The other ECS regions and AMIs are documented here:

http://docs.aws.amazon.com/AmazonECS/latest/developerguide/launch_container_instance.html

Comment (ddollar) [inline dist/staging.json]: Yep. We need to create a Mapping for this. Will add a card.

File: cmd/bootstrap/bin/create (hunk 1 above)
Comment (nzoschke) [inline cmd/bootstrap/bin/create]: Ah, this wasn't obvious to me from the patch. Can you briefly document the design here?

Comment (ddollar) [inline cmd/bootstrap/bin/create]: Done, up in the description.

**Comment by nzoschke** [comment]
:+1:

**Comment by mattmanning** [comment]
:+1:

**Pull request merged**